tests/test_api.py

```diff
@@ -786,3 +786,160 @@ def test_product_components(client, api_path):
     response = client.get(f"{api_path}/components?products={rhel_br.name}")
     assert response.status_code == 200
     assert response.json()["count"] == 1
+
+
+@pytest.mark.django_db(databases=("default", "read_only"), transaction=True)
+def test_component_provides_sources_upstreams(client, api_path):
+    """
+    Taking the following component node structure:
+    /
+    └──(SOURCE) root_comp: top level root component
+        ├──(SOURCE) upstream_comp
+        └──(PROVIDES) dep1_comp
+
+    where upstream is defined as a child node of a root comp and dep1 is also defined as a root component
+    /
+    └──(SOURCE) dep1_comp: top level root component
+            └──(PROVIDES) dep2_comp
+
+    what are the expectations in terms of sources, providers and upstreams on each of these components ?
+
+    +--------------------+---------+----------+-----------+
+    |                    | sources | provides | upstreams |
+    +--------------------+---------+----------+-----------+
+    | root_component     | 0       | 2        | 1         |
+    +--------------------+---------+----------+-----------+
+    | upstream_component | 0       | 0        | 0         |
+    +--------------------+---------+----------+-----------+
+    | dep1_component     | 1       | 1        | 0         |
+    +--------------------+---------+----------+-----------+
+    | dep2_component     | 2       | 0        | 0         |
+    +--------------------+---------+----------+-----------+
+
+    Are the following statements true ?
+
+    * upstream has 0 sources as it is its own SOURCE
+    * should dep1 or dep2 dynamically resolve the top level component's upstreams ?
+    * should dep2 list 2 sources ?
+    * we should not list upstreams in provides right ?
+
+    """
+
+    # create a top level root source component
+    root_comp = ComponentFactory(
+        name="root_comp",
+        type=Component.Type.RPM,
+        arch="src",
+        namespace=Component.Namespace.REDHAT,
+        related_url="https://example.org/related",
+    )
+    root_node, _ = ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=None,
+        purl=root_comp.purl,
+        defaults={"obj": root_comp},
+    )
+
+    # create upstream child node
+    upstream_comp = ComponentFactory(
+        name="upstream_comp",
+        type=Component.Type.PYPI,
```

Comment:
For root components which are source RPMs, the code we have today creates the upstream component with type=RPM and arch=src.
```suggestion
        type=Component.Type.RPM,
        arch="src",
```

EDIT: Oops, I originally put "arch=noarch" here, but that's only for remote-source components and index container images, not source RPMs which should always have "arch=src".